Задания и исполнители

kind-chigik opened the pull request:
---
:mortar_board: [Задания и исполнители](https://up.htmlacademy.ru/yii/1/user/669277/tasks/11)

kind-chigik committed:
Добавлены новые данные для работы с БД

kind-chigik committed:
Добавлены новые данные для работы с БД

kind-chigik committed:
Добавлены роли для таблицы users

kind-chigik committed:
Добавлено 2 новых файла для конвертирование в sql

kind-chigik committed:
Добавлен новый аттрибут роли

В схему БД был новый столбец role. По этой причине модель была обновлена

kind-chigik committed:
Добавлен контролер для обработки страницы юзеров

kind-chigik committed:
Добавлен контролер для обработки страницы задач

kind-chigik committed:
Добавлены шаблоны и виды для страниц задач и юзеров

Ovsyanka reviewed (commented):
Ниже я напишу общие рекомендации, если будут проблемы с применением - в рамках консультации можно обсудить.

## Вьюха

Пойдем с конца, с вьюхи - представим как было бы удобно работать с переданными данными в ней, а затем уже
придумаем как эти данные для нее сформировать.

1. Вьюху [принято начинать](https://www.yiiframework.com/doc/guide/2.0/en/structure-views#creating-views) с описания данных, которые внутри нее доступны.

В твоем случае
```php
/**
 * @var yii\web\View $this
 * @var frontend\models\Users[] $executors
 */
```

Кроме того, что ты просто будешь знать, какие данные в ней доступны, во многих IDE (VSCode в том числе) будет работать
автодополнение.

2. Удобнее работать с объектами, а не массивами. Вместо `$executor['name']` тогда будет `$executor->name`. Писать проще, выглядит
менее загружено символами и автодополнение работает.

Соответственно, меняем во вьюхе везде обращения к элементам массива на обращения к свойствам объектов.

## Контроллер

В контроллере стоит иметь по минимуму логики. (см.
https://www.yiiframework.com/doc/guide/2.0/en/structure-controllers#best-practices). Все эти выборки стоит поместить в
модели, либо другие классы (Repositoryб
например). В твоем случае можелей достаточно.

Как бы нам теперь сделать такой замечательный объект для вьюхи? Ах, я же уже проспойлерил ответ на этот вопрос выше =)
Можно использовать модель пользователя! В ней уже все есть, кроме количества откликов и задач.

Их можно получить так: $executor->getTasks()->count(), а лучше создать метод $executor->getTasksCount(), так как это значение может понадобиться не только на этой странице и хорошо бы потом не копипастить эту цепочку. Если потребуется внести какие-то изменения -это будет гораздо проще сделать. Например, можно так реализовать:

```php
<?php
class Users extends \yii\db\ActiveRecord
{
    private $tasksCount = null;

    public function getTasksCount()
    {
        // this avoid calling a query searching for null primary keys
        if ($this->isNewRecord) {
            return 0;
        }

        if ($this->tasksCount === null) {
            // calculate aggregation on demand from relation
            $this->tasksCount = $this->getTasks()->count();
        }

        return $this->tasksCount;
    }
}
```

Больше об этом можно
почитать в [документации](https://www.yiiframework.com/doc/guide/2.0/en/db-active-record#selecting-extra-fields) yii2.

Получается, в конроллере все что нам надо сделать, это найти всех пользователей, которых мы хотим отобразить и передать
это во вьюху.

Например, так:
```php
<?php
$executors = Users::find()
    ->isExecutor()
    ->sortByRegistrationDate(SORT_DESC)
    ->limit(10)
    ->all();
```

Реализовывается путем определения соответствующих методов в UsersQuery. Вот пара ссылок на тему:
* https://qna.habr.com/q/539531
* https://youtu.be/TyiSYxFL8Gg?t=24m13s

Это всё касается и страницы пользователей и страницы задач.

Ovsyanka reviewed (commented):
view.php файлы сейчас не нужны

Ovsyanka commented on frontend/views/users/index.php:
```diff
@@ -0,0 +1,70 @@
+                <span class="new-task__time">Был на сайте <?= $executor['last_visit']; ?></span>
+            </div>
+            <div class="link-specialization user__search-link--bottom">
+            <?php foreach ($executor['categories_names'] as $category_name): ?>
+                <a href="#" class="link-regular"><?= $category_name; ?></a>
+            <?php endforeach; ?>
```
После рефакторинга, который я описал в ревью, можно заменить на:
```
            <?php foreach ($executor->categories as $category): ?>
                <a href="#" class="link-regular"><?= $category->name; ?></a>
            <?php endforeach; ?>
```

Ovsyanka commented on data/reviews.csv:
```diff
@@ -0,0 +1,12 @@
+review_text,evaluation,task_id,user_id
```
Данные для reviews.csv вообще представлены в replies.csv, если не ошибаюсь, так что стоит replies.csv удалить, так как ты составила свои данные (определенно, более реалистичные :Р) и этот файл больше не будет использоваться. Можешь из него что-нибудь к себе перетащить для объему. На свое усмотрение.

Ovsyanka commented on frontend/models/Users.php:
```diff
@@ -22,6 +22,7 @@
  * @property int|null $rank
  * @property string|null $address
  * @property int $city_id
+ * @property string|null $role
```
Молодец, что не забыла сюда и в остальные места добавить.

kind-chigik committed:
Удалена лишняя роль для пользователя

kind-chigik committed:
Добавлены настройки для основного языка и i18n

kind-chigik committed:
Удален за ненадобностью

Наиболее полные данные по отзывам представлены в другом файле reviews.csv

kind-chigik committed:
Приведен к читаемому виду

Запросы к БД убраны в модели в отдельные методы.

kind-chigik committed:
Добавлены методы, получающие информцию из БД

kind-chigik committed:
Добавлены методы для подсчета кол-ва задач и отзывов пользователей

kind-chigik committed:
Удалены за ненадобностью

kind-chigik committed:
Добавлено описание данных, форматирование, внесены мелкие правки

- Добавлено описание переменных, которые используеются в коде и заголовок страницы
- Данные задач теперь представлены в виде объектов
- Вместо шаблонной иконки теперь выводятся иконки из БД, для каждой задачи своя
- Добавлено форматирование даты

kind-chigik committed:
Удален, т.к. пока не используется

kind-chigik committed:
Удален, т.к. пока не используется

kind-chigik committed:
Добавлено описание данных, форматирование, внесены мелкие правки

- Добавлено описание переменных, которые используеются в коде и заголовок страницы
- Данные задач теперь представлены в виде объектов
- Вместо шаблонных аватаров теперь выводятся аваторы из БД, для каждого пользователя свой. Если аватар не загружен пользователем, по умолчанию устанавливается шаблонный
- Добавлено форматирование даты последнего визита исполнителя на сайт
- Добавлено форматирование текста для блока с количеством задач и отзывов. Задано правильное склонение фраз.

Ovsyanka commented on frontend/models/Users.php:
```diff
@@ -37,6 +37,35 @@
+        }
+
+        if ($this->tasksCount === null) {
+            $this->tasksCount = $this->getTasks0()->count();
```
Я вижу по коду, что Tasks0 отличается от Tasks тем, что Tasks0 это задачи, где пользователь участвует как исполнитель, а Tasks - где как заказчик. Стоит сделать методам "говорящие" имена (критерий Б2. Именование переменных и функций). Например, getExecutorTasksCount, getExecutorTasks.

Ovsyanka commented on frontend/models/Users.php:
```diff
@@ -37,6 +37,35 @@
+    public function getTasksCount()
+    {
+        if ($this->isNewRecord) {
+            return null;
```
Скорее всего ты взяла это из документации к yii2. Несмотря на то, что там указан такой вот вариант с возвращением null, на мой взгляд, уместнее возвращать 0 - ведь если объект пользователя только что создан - он не имел возможности участвовать в заданиях и можно смело утверждать,что их количество равно нулю, а не "не известно". Это важно, поскольку в таком случае у getTasksCount будет точно возвращать int и при работе с ним не надо будет выполнять доп. проверок на null, то есть сложность кода уменьшается.

Ovsyanka reviewed (changes requested):
Прекрасно! Очень все аккуратно, лаконично, понятно. Единственное, вот с task0 промашка получилась.

Еще, ты говорила, что нацелена на выполнение доп. критериев наряду с основными. Многие, скорее всего, у тебя и так выполняются. Но я рекомендую тебе начинаь потихоньку добавлять выполнение тех, которые еще не выполняются, чтобы потом не надо было все сразу делать одновременно. Я бы начал с Д7, на мой взгляд он наиболее полезен и, затем, Д5 - его сильно откладывать не стоит, так как документирование требует практики и чем раньше начнешь - тем лучше. Сначала будет сложно, сильно не запаривайся, а к концу проекта можно будет какие-то неудачные места переписать.

> Д7. Типизация.
> В методах и возвращаемых значениях собственных классов проставлены типы.

> Д5. Документирование методов.
> Все пользовательские функции и методы должны быть задокументированы, используя синтаксис phpDoc.

kind-chigik committed:
Удален лишний путь для удаленного файла replies.csv

kind-chigik committed:
Изменены названия связей и методов на более понятные

kind-chigik committed:
Скорректировано название связи, т.к. оно было изменено в модели User.php

kind-chigik committed:
Скорректировано название свойства для подсчета задач

kind-chigik committed:
Добавлено документирование методов

Ovsyanka commented on frontend/models/Users.php:
```diff
@@ -187,21 +187,21 @@ public function getReviews()
      * @return \yii\db\ActiveQuery|TasksQuery
      */
-    public function getTasks0()
+    public function getExecutorTasks()
```
Тут у тебя в автосгенерированной документации есть подсказка, какого типа возвращаемое значение. Это объект класса TasksQuery.

Ovsyanka commented on src/TaskStrategy.php:
```diff
@@ -87,6 +104,14 @@ public function getNextStatus(string $action) : ?string
+     * 
+     * @return array Массив с объектами действий
+     */
+
```
Пропуск сроки не нужен, ну я вижу, что это опечатка, в остальных местах порядок.

Ovsyanka commented on src/TaskStrategy.php:
```diff
@@ -50,18 +50,35 @@ public function __construct(int $idCustomer, int $idExecutor, string $timeEnd, s
+     * 
+     * @param string $action Действие
+     * 
+     * @return null|string Пустое значение или Статус
```
А в каком случае возвращается пустое значение? _Я бы, кстати, назвал его отсутствием значения или просто null. Пустое значение у меня невольно ассоциируется с чем-то пустым, строкой, например, а это уже подразумевает наличие чего-то, пусть и пустого. Ну это так, информация к размышлению =)_

По сути, эта функция никогда не должна возвращать null - ведь статус для действия всегда должен быть определен, если только это не несуществующее действие. И если он ен определен - это означает, что программист где-то допустил ошибку. Так что уместно возвращать только строку со статусом, а в случае если статус не был определен - исключение класса [LogicException](https://www.php.net/manual/ru/class.logicexception.php):
> Исключение, которое представляет ошибку в логике программы. Такой тип исключений должен непосредственно привести к исправлениям в вашем коде.

Если не ошибаюсь, тут уместен его подкласс [DomainException](https://www.php.net/manual/ru/class.domainexception.php):
> Создается исключение, если значение не соответствует определенным допустимым данным домена.

Заметь, что документация "обнажила" этот момент, который я ранее пропустил =) Документация - это хорошо :Р 

Отмечу, что немного спорный момент насчет того, стоит ли ее писать для более простых методов, особенно если они автосгенерированы. Время на ее написание и поддержку может не окупаться, посколько пользы от нее мало - такие методы зачастую достаточно очевидны. Но в критериях к учебному проекту нет оговорок на этот счет, так что для этого проекта ориентируйся на то, что она нужна ко всем методам.

Ovsyanka reviewed (approved):
Хорошее начало!

Есть пара комментариев, их проработать можно позже.
